---
output: html_document
title: "Interaction effects"
---

```{r setup, include=F}
knitr::opts_chunk$set(echo = FALSE, message = FALSE,
                      fig.path = "../output/interactions/")
suppressMessages(silent <- lapply(
  c("knitr", "kableExtra", "tidyverse", "lme4", "lmerTest", "cowplot", "patchwork"), 
  library, character.only = TRUE))
```

```{r read-sumstats}
phenos <- read_csv("../data/processed/ukb_phenos_longitudinal_unrelated.csv")

fish_crp_gwis <- read_tsv("../data/processed/gwis/oily_fish_hscrp_log_merged_nom")
fish_n3_gwis <- read_tsv("../data/processed/gwis/oily_fish_Omega_3_pct_merged_nom")
n3_crp_gwis <- read_tsv("../data/processed/gwis/Omega_3_pct_hscrp_log_merged_nom")
fish_oil_crp_gwis <- read_tsv("../data/processed/gwis/fish_oil_hscrp_log_merged_nom")
fish_oil_n3_gwis <- read_tsv("../data/processed/gwis/fish_oil_Omega_3_pct_merged_nom")


fish_crp_genes <- read_table("../data/processed/magma/oily_fish_hscrp_log.genes.out") %>%
  arrange(P)
fish_n3_genes <- read_table("../data/processed/magma/oily_fish_Omega_3_pct.genes.out") %>%
  arrange(P)
n3_crp_genes <- read_table("../data/processed/magma/Omega_3_pct_hscrp_log.genes.out") %>%
  arrange(P)
fish_oil_crp_genes <- read_table("../data/processed/magma/fish_oil_hscrp_log.genes.out") %>%
  arrange(P)
fish_oil_n3_genes <- read_table("../data/processed/magma/fish_oil_Omega_3_pct.genes.out") %>%
  arrange(P)
```

The goal of this analysis was to understand the genetic modification of the relationship between dietary omega-3 intake and hsCRP, with plasma omega-3 (as a % of FA from NMR) as an established mediator. We conducted a series of genome-wide interaction studies to explore this possibility, with some of these models used to decompose the N3-hsCRP relationship into components "upstream" and "downstream" of the plasma omega-3 mediator.

# GWIS results for the primary oily fish - hsCRP relationship

```{r prep-qqs}
calc_lambda <- function(x, p=0.5){
  # Calculate genomic inflation lambda value
  x <- x[!is.na(x)]
  x.quantile <- quantile(x, p)
  round(qchisq(1 - x.quantile, 1) / qchisq(p, 1), 2)
}

make_qq <- function(data, pval_col, main=""){
  # Make a quantile-quantile plot
  data <- filter(data, data[[pval_col]] > 0)  # In case extremely low p-values are stored as zero

  # Process p-values
  y_vals <- sort(-log10(data[[pval_col]]))
  x_vals <- -log10(rev(ppoints(length(y_vals))))  # ppoints generates a uniform probability distribution
  lam <- calc_lambda(data[[pval_col]])

  # Trim points at higher p-values (credit to RaMWAS package for code snippet)
  levels = as.integer((x_vals - x_vals[1]) / (tail(x_vals, 1) - x_vals[1]) * 2000)
  keep = c(TRUE, diff(levels) != 0)
  levels = as.integer((y_vals - y_vals[1])/(tail(y_vals, 1) - y_vals[1]) * 2000)
  keep = keep | c(TRUE, diff(levels) != 0)
  keep = which(keep)

  rm(data)  # Deal with memory issues
    
  tibble(x=x_vals[keep], y=y_vals[keep]) %>%
    ggplot(aes(x=x, y=y)) + 
    geom_point() +
    geom_abline(slope=1, linetype="dashed") +
    annotate("text", 
#              label=bquote(lambda == .(lam)),  # NOTE: somehow this bquote call results in the warning about "is.na() applied to type language"
             label=paste("lambda == ", lam), parse=TRUE,
             x=-Inf, y=Inf, hjust=-0.5, vjust=3, size=5) +
    labs(x=expression(-log[10](italic(p)) * " (Expected)"),
         y=expression(-log[10](italic(p)) * " (Observed)"),
         title=main) +
    theme_bw() +
    theme(panel.grid=element_blank())
}
```

```{r prep-manhattans}
make_manhattan <- function(data, 
                           pval_col, snp_col = "SNP", chr_col = "CHR", pos_col = "POS", 
                           threshold = 5e-8, ylims = NULL, main = "",
                           add_labels = TRUE) {

  mh_data <- data %>%
    dplyr::rename(all_of(c(SNP = snp_col, CHR = chr_col, POS = pos_col, P = pval_col))) %>%
    filter(!is.na(P)) %>%
    mutate(P = as.numeric(P),
           P = ifelse(P == 0, min(1e-300, min(P[P != 0], na.rm = TRUE)), P),  # Remove P = 0
           nlp = -log10(P))
  
  # Trim points in crowded regions (credit to RaMWAS package for code snippet)
  yfac = as.integer(mh_data$nlp * 100) + 1L
  yorder = sort.list(yfac)
  yfac <- factor(yfac, levels = as.character(seq_len(max(yfac))))
  ygroup <- split(seq_along(yfac), yfac)
  for (i in seq_along(ygroup)) {
    if (length(ygroup[[i]]) > 300) {
      ygroup[[i]] <- sample(ygroup[[i]], size = 300, replace = FALSE)
    }
  }
  keep <- unlist(ygroup, use.names = FALSE)
  
  mh_data <- mh_data %>%
    select(SNP, CHR, POS, nlp) %>%
    dplyr::slice(keep) %>%
    mutate(POS = as.numeric(as.character(POS)),
           CHR = factor(CHR, levels = c(1:22, "X"))) %>%
    arrange(CHR, POS) %>%
    mutate(pos_idx = seq(1, nrow(.)))
  
  suppressWarnings(chr_lengths <- sapply(c(1:22, "X"), function(chr) {
    with(mh_data, max(POS[CHR == chr], na.rm = TRUE))
  }))
  chr_lengths <- ifelse(is.infinite(chr_lengths), 0, chr_lengths)
  chr_start_pos <- cumsum(chr_lengths) - chr_lengths
  
  mh_data <- mh_data %>%
    mutate(x_coord = chr_start_pos[CHR] + POS,
           color = CHR,
    ) %>%
    arrange(as.integer(color), nlp) #%>%
#     distinct(SNP, .keep_all = T)
  
  lims <- mh_data %>%
    group_by(CHR) %>%
    summarise(avg_coord = (min(x_coord) + max(x_coord)) / 2)
  
  newcols <- setNames(
    rep(x = c("#AAAAAA", "#8A8A8A"), length.out = 23),  # Gray/dark gray for alternating chromosomes
    levels(factor(lims$CHR))
  )
  
  mh_plt <- ggplot() +
    geom_point(data = mh_data, 
               aes(x = x_coord, y = nlp, color = factor(color)), 
               size = 0.75, alpha = 1) +
    geom_hline(yintercept = -log10(threshold), linetype = "dashed", color = "black") + 
    scale_x_continuous(breaks = lims$avg_coord[c(1:16, 18, 20, 20, 22, 23)], 
                       labels = c(1:16, 18, 20, 20, 22, "X"), 
                       expand = c(0,0)) +
    scale_y_continuous(name = expression(-log[10](italic(p)))) +
    scale_colour_manual(values = newcols, guide = "none") +
    scale_fill_manual(name = "Color", values = newcols) +
    labs(title = main) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(vjust = -1.5),
          panel.grid.minor = element_blank(),
          panel.grid.major.x = element_blank())
  if (!is.null(ylims)) mh_plt <- mh_plt + coord_cartesian(ylim = ylims)
  if (add_labels) mh_plt <- mh_plt +
    geom_text(data = filter(mh_data, nlp > -log10(threshold)),
               aes(x = x_coord, y = nlp, label = SNP))
  
  mh_plt
}

gene_bonferroni <- 0.05 / nrow(fish_crp_genes)
```

```{r primary-manhattans}
make_manhattan(fish_crp_gwis, "P_int", threshold = 5e-8) +
  labs(title = "Variant-level: oily fish - hsCRP")

include_graphics("../data/processed/gwis/qq_plots/oily_fish_hscrp_log_robust_QQ.pdf")

make_manhattan(fish_crp_gwis, "P_joint", threshold = 5e-8, add_labels = FALSE) +
  labs(title = "Variant-level joint test: oily fish - hsCRP")

make_manhattan(fish_crp_genes, "P", snp_col = "GENE", pos_col = "START",
               threshold = gene_bonferroni) +
  labs(title = "Gene-level: oily fish - hsCRP")
```

# GWIS results using the plasma omega-3 mediator

We can then explore GWIS results informed by the plasma omega-3 mediator. These corresponding to either "upstream" (genetic modifiers of the fish - plasma N3 relationship) or "downstream" (genetic modifiers of the plasma N3 - hsCRP relationship). These models provide both additional biological insights and may increase statistical power through greater precision (i.e., plasma N3 as a better proxy for true dietary omega-3 intake than self report). 

```{r mediated-manhattans}
make_manhattan(fish_n3_gwis, "P_int", threshold = 5e-8) +
  labs(title = "Variant-level: oily fish - plasma omega-3 (\"upstream\")")
include_graphics("../data/processed/gwis/qq_plots/fish_oil_Omega_3_pct_robust_QQ.pdf")
make_manhattan(fish_crp_gwis, "P_joint", threshold = 5e-8, add_labels = FALSE) +
  labs(title = "Variant-level joint test: oily fish - plasma omega-3 (\"upstream\")")
make_manhattan(fish_n3_genes, "P", snp_col = "SYMBOL", pos_col = "START",
               threshold = gene_bonferroni) +
  labs(title = "Gene-level: oily fish - plasma omega-3 (\"upstream\")")

make_manhattan(n3_crp_gwis, "P_int", threshold = 5e-8) +
  labs(title = "Variant-level: plasma omega-3 - hsCRP (\"downstream\")")
include_graphics("../data/processed/gwis/qq_plots/Omega_3_pct_hscrp_log_robust_QQ.pdf")
make_manhattan(fish_crp_gwis, "P_joint", threshold = 5e-8, add_labels = FALSE) +
  labs(title = "Variant-level joint test: plasma omega-3 - hsCRP (\"downstream\")")
make_manhattan(n3_crp_genes, "P", snp_col = "SYMBOL", pos_col = "START",
               threshold = gene_bonferroni) +
  labs(title = "Gene-level: plasma omega-3 - hsCRP (\"downstream\")")
```

# GWIS results using fish oil instead of oily fish intake

Fish oil is an alternative source of dietary omega-3 intake that, while not highly correlated with fish intake, might be expected to show similar genetic architecture of response modification. We will look at primary and "upstream" GWIS results from models replacing oily fish with fish oil. 

```{r fish-oil-manhattans}
make_manhattan(fish_oil_crp_gwis, "P_int", threshold = 5e-8) +
  labs(title = "Variant-level: fish oil - hsCRP")
make_manhattan(fish_oil_crp_genes, "P", snp_col = "SYMBOL", pos_col = "START",
               threshold = gene_bonferroni) +
  labs(title = "Gene-level: fish oil - hsCRP")

make_manhattan(fish_oil_n3_gwis, "P_int", threshold = 5e-8) +
  labs(title = "Variant-level: fish oil - plasma omega-3 (\"upstream\")")
make_manhattan(fish_oil_n3_genes, "P", snp_col = "SYMBOL", pos_col = "START",
               threshold = gene_bonferroni) +
  labs(title = "Gene-level: fish oil - plasma omega-3 (\"upstream\")")
```

Furthermore, a genetic correlation between (G x Fish) and (G x Fish oil) would strengthen the mechanistic basis for these conclusions about interactions. Do we observe this in practice?

```{r genetic-correlations}
hscrp_gc <- read_table("../data/processed/ldsc/oily_fish_fish_oil_hscrp_log.log", 
                       skip = 60, n_max = 1)
n3_gc <- read_table("../data/processed/ldsc/oily_fish_fish_oil_Omega_3_pct.log", 
                       skip = 60, n_max = 1)

bind_rows(list(
  hscrp_gc,
  n3_gc
)) %>%
  select(p1, p2, rg, p) %>%
  mutate(across(c(p1, p2), basename),
         across(c(rg, p), ~ signif(., 2)))
```

We don't see any significant genetic correlation for the interaction estimates in this case.

# Gene-specific follow-up

```{r prep-gene-followup}
run_gxe <- function(g, e, y, 
                    covars = c("age", "sex"), 
                    df = phenos,
                    std = FALSE) {
  if (std) df <- mutate(df, across(all_of(c(e, y)), ~ scale(.)))
  lm_str <- paste0(y, " ~ ", g, " * ", e, " + ", paste0(covars, collapse = " + "))
  lm_summ <- tryCatch({
    lm_fit <- lm(as.formula(lm_str), data = df) 
    lm_fit %>%
      broom::tidy() %>%
      filter(term == paste0(g, ":", e)) %>%
      mutate(residual_df = lm_fit$df.residual)
  }, error = tibble(NA))
  lm_summ
}

make_gene_snp_plot <- function(ss_df, p_col = "robust_P_Value_Interaction") {
  ss_df %>%
    mutate(nlp_int = -log10(!!sym(p_col))) %>%
    ggplot(aes(x = POS, y = nlp_int)) +
    geom_point() +
    geom_text(aes(label = RSID)) +
    labs(y=expression(-log[10](p[int])))
}

make_int_effect_plot <- function(all_maf_ss_df, e, top_rsid) {
  beta_vec <- all_maf_ss_df[[paste0("Beta_G-", e)]]
  ylims <- median(beta_vec, na.rm = TRUE) + c(-5, 5) * mad(beta_vec, na.rm = TRUE)
  all_maf_ss_df %>%
    mutate(int_estimate = !!sym(paste0("Beta_G-", e)),
           int_se = !!sym(paste0("SE_Beta_G-", e))) %>%
    arrange(int_estimate) %>%
    mutate(l95 = int_estimate - 1.96 * int_se,
           u95 = int_estimate + 1.96 * int_se,
           top_gwis = (RSID == gsub("_.*", "", top_rsid)),
           SNPID = factor(SNPID, levels = SNPID)) %>%
    ggplot(aes(x = SNPID, y = int_estimate)) +
    geom_point(aes(color = top_gwis)) +
    geom_errorbar(aes(ymin=l95, ymax=u95, color = top_gwis), width = 0.2) +
    geom_hline(yintercept = 0, color = "gray") +
    # geom_text(aes(label = RSID)) +
    scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
                       breaks = c(TRUE),
                       labels = c("Top variant from GWIS")) +
    labs(x = "rsID", y = "Interaction effect estimate (95% CI)") +
    coord_cartesian(ylim = ylims) +
    theme(axis.text.x = element_blank(),
          # axis.text.x = element_text(angle = 45, hjust = 0.9),
          legend.position = "bottom", legend.title = element_blank(), 
          legend.key = element_blank())
}

make_int_effect_maf_plot <- function(all_maf_ss_df, e, top_rsid) {
  all_maf_ss_df %>%
    mutate(int_estimate = !!sym(paste0("Beta_G-", e)),
           int_se = !!sym(paste0("SE_Beta_G-", e))) %>%
    mutate(l95 = int_estimate - 1.96 * int_se,
           u95 = int_estimate + 1.96 * int_se,
           top_gwis = (RSID == gsub("_.*", "", top_rsid)),
           MAF = pmin(AF, 1 - AF)) %>%
    arrange(top_gwis) %>%
    ggplot(aes(x = MAF, y = int_estimate)) +
    geom_point(aes(color = top_gwis)) +
    geom_errorbar(aes(color = top_gwis, ymin=l95, ymax=u95), width = 0) +
    geom_hline(yintercept = 0, color="gray") +
    scale_x_log10() +
    scale_color_manual(values = c(`FALSE` = "black", `TRUE` = "red"), 
                       breaks = c(TRUE),
                       labels = c("Top variant from GWIS")) +
    labs(x = "MAF", y = "Interaction effect estimate (95% CI)") +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
          legend.position = "bottom", legend.title = element_blank(), 
          legend.key = element_blank())
}

make_int_effect_cond_plot <- function(ss_df, ss_cond_df, e, top_rsid) {
  beta_vec <- ss_cond_df[[paste0("Beta_G-", e)]]
  ylims <- median(beta_vec, na.rm = TRUE) + c(-5, 5) * mad(beta_vec, na.rm = TRUE)
  bind_rows(list(unconditional = ss_df, conditional = ss_cond_df), 
            .id = "adjustment") %>%
    mutate(int_estimate = !!sym(paste0("Beta_G-", e)),
           int_se = !!sym(paste0("SE_Beta_G-", e))) %>%
    select(SNPID, RSID, AF, adjustment, int_estimate, int_se) %>%
    arrange(int_estimate) %>%
    mutate(l95 = int_estimate - 1.96 * int_se,
           u95 = int_estimate + 1.96 * int_se,
           top_gwis = (RSID == gsub("_.*", "", top_rsid)),
           MAF = pmin(AF, 1 - AF),
           SNPID = factor(SNPID, levels = unique(SNPID[adjustment == "unconditional"]))) %>%
    filter(!top_gwis) %>%
    ggplot(aes(x = SNPID, y = int_estimate, color = adjustment)) +
    geom_point(position = position_dodge(width = 0.1)) +
    geom_errorbar(aes(ymin = l95, ymax = u95), width = 0, 
                  position = position_dodge(width = 0.1)) +
    geom_hline(yintercept = 0, color = "gray") + 
    labs(x = "Interaction effect (95% CI) - unconditional", 
         y = "Interaction effect (95% CI) - conditional on top variant") +
    coord_cartesian(ylim = ylims) +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
          legend.position = "bottom", legend.title = element_blank(), 
          legend.key = element_blank())
}

make_int_p_cond_plot <- function(ss_df, ss_cond_df, e, top_rsid) {
  bind_rows(list(unconditional = ss_df, conditional = ss_cond_df), 
            .id = "adjustment") %>%
    mutate(int_nlp = -log10(P_Value_Interaction)) %>%
    select(SNPID, RSID, AF, adjustment, int_nlp) %>%
    pivot_wider(names_from = "adjustment", values_from = "int_nlp") %>%
    mutate(top_gwis = (RSID == gsub("_.*", "", top_rsid)),
           MAF = pmin(AF, 1 - AF)) %>%
    filter(!top_gwis) %>%
    ggplot(aes(x = unconditional, y = conditional, color = MAF)) +
    geom_point() +
    geom_abline(slope = 1, intercept = 0, color="gray") +
    scale_color_continuous(trans = "log", breaks = c(0.001, 0.01, 0.1)) +
    labs(x = "Interaction p-value (unconditional)", 
         y = "Interaction p-value (conditional on top variant)") +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.9),
          legend.position = "bottom")
}

run_gene_follow_up <- function(gene, e, y, base_covars = covar_sets$ffqAdj,
                               snp_idx = 1) {
  e_regression <- if (e == "fish_oil") "fish_oil_touchscreen" else e
  
  gwis_sumstats_df <- read_tsv(paste0("../data/processed/gene_followup/", 
                                      e, "_", y, "_", gene, "_subset"))
  gene_snp_plt <- make_gene_snp_plot(gwis_sumstats_df)
  print(gene_snp_plt)
  
  top_rsid_df <- read_tsv(paste0("../data/processed/gene_followup/", gene, 
                                 "_top_rsid.raw")) %>%
    select(id = IID, matches("^rs")) %>%
    mutate(id = as.character(id))
  top_rsid <- names(top_rsid_df)[snp_idx + 1]
  regression_df <- inner_join(phenos, top_rsid_df, by = "id")
  
  sensitivity_covar_list <- list(
    primary = base_covars,
    add_E_by_gPC = c(base_covars, paste0(e_regression, " * gPC", 1:10)),
    add_G_by_Cov = paste0(top_rsid, " * ", base_covars)
  )
  sensitivity_covar_res_df <- tibble(covar_set = names(sensitivity_covar_list)) %>%
    rowwise() %>%
    mutate(model_res = list(run_gxe(top_rsid, e_regression, y,
                                    sensitivity_covar_list[[covar_set]], 
                                    regression_df, std = TRUE))) %>%
    unnest(model_res) %>%
    mutate(across(estimate:p.value, ~ signif(., 3)))
  print(sensitivity_covar_res_df)
  
  sensitivity_e_list <- list(
    primary = e_regression,
    oily_fish = "oily_fish",
    nonoily_fish = "nonoily_fish",
    N3FA = "N3FA",
    fish_oil_touchscreen = "fish_oil_touchscreen"
  )
  sensitivity_e_res_df <- tibble(e_test = names(sensitivity_e_list)) %>%
    rowwise() %>%
    mutate(model_res = list(run_gxe(top_rsid, sensitivity_e_list[[e_test]], y,
                                    base_covars, 
                                    regression_df, std = TRUE))) %>%
    unnest(model_res) %>%
    mutate(across(estimate:p.value, ~ signif(., 3)))
  print(sensitivity_e_res_df)
  
  all_maf_sumstats_df <- read_tsv(paste0("../data/processed/gene_followup/", 
                                         e, "_", y, "_", gene, "_regressions"))
  all_maf_cond_sumstats_df <- read_tsv(paste0("../data/processed/gene_followup/", 
                                              e, "_", y, "_", gene, "_regressions_cond"))
  int_effect_plt <- make_int_effect_plot(all_maf_sumstats_df, e, top_rsid)
  int_effect_maf_plt <- make_int_effect_maf_plot(all_maf_sumstats_df, e, top_rsid)
  int_effect_cond_maf_plt <- make_int_effect_maf_plot(
    filter(all_maf_cond_sumstats_df, RSID != gsub("_.*", "", top_rsid)), 
    e, top_rsid
  )
  int_effect_cond_plt <- make_int_effect_cond_plot(all_maf_sumstats_df,
                                                   all_maf_cond_sumstats_df,
                                                   e, top_rsid)
  int_p_cond_plt <- make_int_p_cond_plot(all_maf_sumstats_df, 
                                         all_maf_cond_sumstats_df,
                                         e, top_rsid)
  print(int_effect_plt)
  print(int_effect_maf_plt)
  print(int_effect_cond_maf_plt)
  print(int_effect_cond_plt)
  print(int_p_cond_plt)
}

phenos <- read_csv("../data/processed/ukb_phenos_longitudinal_unrelated.csv") %>%
  filter(instance == 0) %>%
  mutate(id = as.character(id))

minimal_covars <- c("age", "age_squared", "sex", "ageBySex")
ses_hl_covars <- c("ac", "income", "education", "smoking", "alcohol")
ffq_covars <- c("cooked_veg", "raw_veg", "fresh_fruit", "prmeat", "whole_bread")
covar_sets <- list(
  minimal = minimal_covars,
  adj = c(minimal_covars, ses_hl_covars),
  mdsAdj = c(minimal_covars, ses_hl_covars, "mds"),
  ffqAdj = c(minimal_covars, ses_hl_covars, ffq_covars)
)
```

## rs149473 (tentatively assigned to *LINC00170*)

```{r rs149473-sumstats}
run_gene_follow_up("parp8", "oily_fish", "Omega_3_pct", snp_idx = 2)
```

## *FADS* cluster (FADS1, FADS2) (oily fish -> N3 %)

```{r fads-cluster-sumstats}
run_gene_follow_up("fads1", "oily_fish", "Omega_3_pct")
run_gene_follow_up("fads2", "oily_fish", "Omega_3_pct")
# run_gene_follow_up("tmem258", "oily_fish", "Omega_3_pct")

# gene <- "fads1"
# e <- "oily_fish"
# y <- "Omega_3_pct"
# base_covars <- covar_sets$ffqAdj
# snp_idx <- 1
# top_rsid_df <- read_tsv(paste0("../data/processed/gene_followup/", gene, 
#                                "_top_rsid.raw")) %>%
#   select(id = IID, matches("^rs")) %>%
#   mutate(id = as.character(id))
# second_rsid_df <- read_tsv("fads1_second_variant.raw") %>%
#   select(id = IID, matches("^rs")) %>%
#   mutate(id = as.character(id))
# regression_df <- phenos %>%
#   inner_join(top_rsid_df, by = "id") %>%
#   inner_join(second_rsid_df, by = "id") %>%
#   mutate(gxe_covar = oily_fish * rs174550_T)
# run_gxe("rs174550_T", e, y, base_covars, regression_df, std = TRUE)
# run_gxe("rs174549_G", e, y, base_covars, regression_df, std = TRUE)
# run_gxe("rs174549_G", e, y, c(base_covars, "rs174550_T", "gxe_covar"), regression_df, std = TRUE)
```

## *SPRED1* (N3 % -> hsCRP)

```{r spred1-sumstats}
run_gene_follow_up("spred1", "Omega_3_pct", "hscrp_log")
```

## *MICA* (N3 % -> hsCRP)

```{r mica-sumstats}
run_gene_follow_up("mica", "Omega_3_pct", "hscrp_log")
```

## *INPP4A* (fish oil -> N3 %)

```{r inpp4a-sumstats}
run_gene_follow_up("inpp4a", "fish_oil", "Omega_3_pct")
```

## *SPP1* (fish oil -> N3 %)

```{r spp1-sumstats}
run_gene_follow_up("spp1", "fish_oil", "Omega_3_pct")
```

```{r tmem258-regressions, eval=F}
extract_gt_vec <- function()
a <- vcfR::read.vcfR("../data/processed/genotypes/tmem258_genotypes.vcf.gz")

var_info <- vcfR::getFIX(a)
gt <- vcfR::extract.gt(a, element = "DS", 
                       mask = (var_info[, "ID"] == "rs102274"), 
                       as.numeric = TRUE)
gt_df <- gt %>%
  as_tibble(rownames = "id") %>%
  rename(rs102274 = value)

phenos <- left_join(phenos, gt_df, by = "id")

run_gxe("rs102274", "oily_fish", "Omega_3_pct", covar_sets$ffqAdj, phenos)
run_gxe("rs102274", "oily_fish", "Omega_3_pct", 
        paste0(covar_sets$ffqAdj, " * rs102274"), 
        phenos)
run_gxe("rs102274", "oily_fish", "Omega_3_pct", 
        c(covar_sets$ffqAdj, paste0(paste0("gPC", 1:5), " * oily_fish")), 
        phenos)

# phenos %>%
#   filter(!is.na(oily_fish), !is.na(hscrp_log)) %>%
#   mutate(fish_tert = cut(oily_fish, c(-1, 0, 0.2, 10), include.lowest=FALSE,
#                          labels=c("None", "<=1 serving/wk", ">1 serving/wk")),
#          rs622087_int = factor(round(rs622087), labels=c("TT", "GT", "GG"))) %>%
#   group_by(fish_tert, rs622087_int) %>%
#   summarise(m = mean(hscrp_log),
#             se = sd(hscrp_log) / sqrt(n()),
#             l95 = m - 1.96 * se,
#             u95 = m + 1.96 * se) %>%
#   ggplot(aes(x=rs622087_int, y=m, color=fish_tert)) +
#   geom_point(position=position_dodge(width=0.2)) +
#   geom_errorbar(aes(ymin=l95, ymax=u95), position=position_dodge(width=0.2),
#                 width=0.2) +
#   scale_color_discrete(name="Oily fish intake (FFQ)") +
#   labs(x="rs622087",
#        y="Mean log(hsCRP) w/ 95% CI")
```


# Power calculations for replication

```{r power-calcs}
library(ESPRESSO.GxE)

param_grid <- expand_grid(
  maf = c(0.1, 0.25),
  gxe_beta = seq(0.01, 0.04, by = 0.005),
  N = c(1500, 25000),
  env_reliability = 1
)

simulation_params <- tibble(
  scenario.id = 1:nrow(param_grid),
  seed.val = 1,
  numsims = 100,
  numcases = 1000,  # These are just to prevent errors - shouldn't affect results
  numcontrols = 1000,  #
  numsubjects = param_grid$N,
  interaction.OR = 1,  #
  interaction.efkt = param_grid$gxe_beta,
  p.val = 0.05,
  power = 0.8
)

pheno_params <- tibble(
  scenario.id = 1:nrow(param_grid),
  pheno.model = 1,
  disease.prev = 0.1,
  pheno.mean = 0, 
  pheno.sd = 1,
  pheno.sensitivity = 1,   # These are just to prevent errors - shouldn't affect results
  pheno.specificity = 1,   #
  pheno.reliability = 1
)

# Genotype parameters
geno_params <- tibble(
  scenario.id = 1:nrow(param_grid),
  geno.model = 1,
  MAF = param_grid$maf,
  geno.efkt = 0.01,
  geno.sensitivity = 1,
  geno.specificity = 1
)

# Environment parameters
env_params <- data.frame(
  scenario.id = 1:nrow(param_grid),
  env.model = 1,
  env.prevalence = 0.1,
  env.efkt = 0.2,
  env.mean = 0,
  env.sd = 1,
  env.low.lim = 0,
  env.up.lim = 1,
  env.reliability = param_grid$env_reliability
)

run_simulations <- function(sp, pp, gp, ep, s2r) {
    system("rm output.csv")  # Just to be sure (since ESPRESSO.GxE doesn't overwrite)
    capture.output(run.espresso.GxE(sp, pp, gp, ep, s2r))
    results <- suppressWarnings(read_delim("output.csv", delim = ";"))
    new_colnames <- c(names(results)[1:34], 1:4, names(results)[35:38])
    results_fmt <- read_delim("output.csv", delim = ";", 
                              skip = 1, col_names = FALSE) %>%
      setNames(new_colnames) %>%
      select(numsubjects, interaction.efkt, MAF, env.reliability,
             empirical.power)
}

sim_res <- run_simulations(simulation_params, pheno_params, geno_params, 
                           env_params, 1:nrow(param_grid))

sim_res %>%
  mutate(MAF = factor(MAF)) %>%
  ggplot(aes(x = interaction.efkt, y = empirical.power, 
             group = MAF, color = MAF)) +
  geom_point() +
  geom_line() +
  geom_vline(xintercept = 0.025, linetype = "dotted", color = "black") +
  geom_vline(xintercept = 0.035, linetype = "dotted", color = "black") +
  facet_wrap(vars(numsubjects), nrow = 1, 
             labeller = labeller(numsubjects = c("1500" = "N = 1500 (w/ LC-MS)", 
                                                 "25000"="N = 25000 (whole cohort)"))) +
  labs(x = expression("Interaction effect (" * SD[y] * " / " * SD[e] * " / allele)"), 
       y = expression("Empirical power (" * italic(p) * " < 0.05)"))
```

# Archive

## AHA EPI Lifestyle abstract

```{r aha-results}
fads1_var_annot <- read_table("../data/processed/annovar/fads1_variants.variant_function",
                              col_names = c("annot", "gene", "CHR", "POS", "POS2", "REF", "ALT", "RSID"))
fads1_cond_regressions <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads1_regressions_cond") %>%
  left_join(select(fads1_var_annot, CHR, POS, annot), by = c("CHR", "POS"))
fads1_regressions <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads1_regressions") %>%
  left_join(select(fads1_var_annot, CHR, POS, annot), by = c("CHR", "POS"))
nonstd_effect <- fads1_cond_regressions %>%
  filter(AF < 0.01) %>%
  arrange(P_Value_Interaction) %>%
  slice(1) %>%
  select(`Beta_G-oily_fish`)
std_effect <- nonstd_effect$`Beta_G-oily_fish` / sd(phenos$Omega_3_pct, na.rm = TRUE) * 
  sd(phenos$oily_fish, na.rm = TRUE)
                                                           

fads2_var_annot <- read_table("../data/processed/annovar/fads2_variants.variant_function",
                              col_names = c("annot", "gene", "CHR", "POS", "POS2", "REF", "ALT", "RSID"))
fads2_cond_regressions <- read_tsv("../data/processed/gene_followup/oily_fish_Omega_3_pct_fads2_regressions_cond") %>%
  left_join(select(fads2_var_annot, CHR, POS, annot), by = c("CHR", "POS"))
```

